### PR TITLE
Don't allow the hoisting of GT_CLS_VARs that were assigned a constant value.

### PR DESCRIPTION
**Pull request by briansull**
Fixes Issue #26417

### CODE REVIEW

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL
@BruceForstall

File: src/jit/optimizer.cpp [hunk 1]
```diff
@@ -6789,6 +6789,12 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                 return false;
             }
 
+            // Don't hoist constants or nodes that have been assigned a constant value.
+            if (m_compiler->vnStore->IsVNConstant(node->gtVNPair.GetLiberal()))
```
Comment (CarolEidt) [inline src/jit/optimizer.cpp]: At a minimum this needs a much better comment, as it is unclear *why* it should not hoist constants.
But I wonder whether this is excessively conservative. Have you run diffs? @mikedn said here: https://github.com/dotnet/coreclr/issues/26417#issuecomment-525884714 that perhaps it should be checking the Conservative VN for this case.

Comment (briansull) [reply src/jit/optimizer.cpp]: Nothing good happens by placing hoisting a constant into the loop pre-header here.

The CSE logic currently has code to avoid performing  CSE's on constants.
Unfortunately it checks the Conservative VN, which allows this bug to creep in:

optcse.cpp line 649
```
                // Don't CSE constant values, instead let the Value Number
                // based Assertion Prop phase handle them.  Here, unlike
                // the rest of optCSE, we use the conservative value number
                // rather than the liberal one, since the conservative one
                // is what the Value Number based Assertion Prop will use
                // and the point is to avoid optimizing cases that it will
                // handle.
                //
                if (vnStore->IsVNConstant(vnStore->VNConservativeNormalValue(tree->gtVNPair)))
                {
                    continue;
                }
```

Comment (briansull) [reply src/jit/optimizer.cpp]: We will still hoist invariant expression that contain constant values, just not no a solitary constant, or in the case of this bug, a GT_CLS_VAR that has been assigned a constant value.
For hoisting we are simply looking at the Liberal Value number has hoisting it  if it identifies as a constant.

**Comment by briansull** [comment]
Updated fix.

@dotnet/jit-contrib PTAL

File: src/jit/optimizer.cpp [hunk 2]
```diff
@@ -6795,8 +6795,18 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
...
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // Don't allow the hoisting of GT_CLS_VAR an GT_LCL_VAR that were assigned a constant value.
+                // But allow constant GT_ARGPLACE nodes which occur when we have constant args.
```
Comment (CarolEidt) [inline src/jit/optimizer.cpp]: I might just request a bit more lengthy comment here. As I understand it, the reason we can't do this thing that seems safe, though perhaps not all that useful, is that we're not checking safety that something that's known to evaluate to a constant is not being modified within the loop. If I have that right, it would be nice to add that here so that people (like me) won't be confused in future when they see this.

Comment (erozenfeld) [reply src/jit/optimizer.cpp]: Is this the right place for this fix? optVNIsLoopInvariant has logic for determining whether the definition is inside the loop. That logic is not executed in this case because of this code:

https://github.com/dotnet/coreclr/blob/ef585a4658a830945f24d635b234f44bd0c81e84/src/jit/optimizer.cpp#L7197-L7201

Perhaps that short-circuit should be removed so that we try to determine if the definition is inside the loop even for const vn's.

File: src/jit/optimizer.cpp [hunk 3]
```diff
@@ -6795,8 +6795,18 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
 
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // Don't allow the hoisting of GT_CLS_VAR an GT_LCL_VAR that were assigned a constant value.
```
Comment (erozenfeld) [inline src/jit/optimizer.cpp]: This comment is not accurate: we are not allowing any nodes (not just GT_CLS_VAR and GT_LCL_VAR) except GT_ARGPLACE.

**Comment by erozenfeld** [comment]
What are the diffs with this fix?

File: src/jit/optimizer.cpp (hunk 2 above)
Comment (briansull) [reply src/jit/optimizer.cpp]: That function examines the `vn` passed in and sees it it is a  VNF_PhiDef or a VNF_PhiMemoryDef.

So it can only identify the GT_LCL_VARS that have definitions that are tracked by Phi nodes.

When we have a constant `vn` we can't trace back to the definition.

File: src/jit/optimizer.cpp (hunk 3 above)
Comment (briansull) [reply src/jit/optimizer.cpp]: I will try to clarify the comment.  But yes, the intention is to restrict the nodes that can be short-circuit later to just be plain constant nodes or the GT_ARGPLACE nodes.

Comment (briansull) [reply src/jit/optimizer.cpp]: Updated to:

```
            if (m_compiler->vnStore->IsVNConstant(vn))
            {
                // We only allow true constant nodes  (i.e. GT_CNS_INT, etc...) or a GT_ARGPLACE nodes.
                // It is unsafe to allow a GT_CLS_VAR that has been assigned a constant.
                // The logic in optVNIsLoopInvariant woudl consider it to be loop-invariant, even
                // if the assignment of the constant to the GT_CLS_VAR was inside the loop.
                //
                if (!tree->OperIsConst() && tree->OperGet() != GT_ARGPLACE)
                {
                    return false;
                }
            }
```

File: src/jit/optimizer.cpp (hunk 2 above)
Comment (erozenfeld) [reply src/jit/optimizer.cpp]: optVNIsLoopInvariant also has logic for when the vn is not VNF_PhiDef or a VNF_PhiMemoryDef: it checks function args if vn is a function and it checks vn loop annotations if vn is not a function.

Comment (briansull) [reply src/jit/optimizer.cpp]: That handles things like binary operators such as Add, Sub, etc..
Constant values are terminal node with no children, only a value.

**Comment by briansull** [comment]
In corelib diffs, one method had a diff

jit-diff diff --pmi --base D:\fxkit\coreclr\bin\Product\Windows_NT.x64.Checked --diff D:\fxkit\baseline\coreclr\bin\Product\Windows_NT.x64.Checked --core_root D:\fxkit\coreclr\bin\tests\Windows_NT.x64.Release\Tests\Core_Root --corelib --tsv
Using --output d:\fxkit\baseline\coreclr\bin\diffs
Beginning PMI Diffs for System.Private.CoreLib.dll
| Finished 1/1 Base 1/1 Diff [310.7 sec]
Completed PMI Diffs for System.Private.CoreLib.dll in 310.79s
Diffs (if any) can be viewed by comparing: d:\fxkit\baseline\coreclr\bin\diffs\dasmset_5\base d:\fxkit\baseline\coreclr\bin\diffs\dasmset_5\diff
Analyzing diffs...
Found 1 files with textual diffs.
PMI Diffs for System.Private.CoreLib.dll for  default jit
Summary:
(Lower is better)
Total bytes of diff: 129 (0.00% of base)
    diff is a regression.
Top file regressions by size (bytes):
         129 : System.Private.CoreLib.dasm (0.00% of base)
1 total files with size differences (0 improved, 1 regressed), 0 unchanged.
Top method regressions by size (bytes):
         129 (25.05% of base) : System.Private.CoreLib.dasm - SyncTextWriter:WriteLine(ref):this (4 base, 5 diff methods)
Top method regressions by size (percentage):
         129 (25.05% of base) : System.Private.CoreLib.dasm - SyncTextWriter:WriteLine(ref):this (4 base, 5 diff methods)
1 total methods with size differences (0 improved, 1 regressed), 20322 unchanged.
Completed analysis in 10.57s

File: src/jit/optimizer.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/optimizer.cpp]: > We will still hoist invariant expressions that contain constant values, 

This is no longer the case with the updated fix, right?

File: src/jit/optimizer.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/optimizer.cpp]: > That handles things like binary operators such as Add, Sub, etc..

That part of VN invariance is actually dubious since it's redundant to what the rest of loop hoisting does.

File: src/jit/optimizer.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/optimizer.cpp]: Ah, I see that there was another updated that deals with this.

File: src/jit/optimizer.cpp (hunk 2 above)
Comment (briansull) [reply src/jit/optimizer.cpp]: One important case is handled here:
We have several "pure" Jit helper functions that take constants as arguments.
It is important that such calls get hoisted into the loop pre-header and then turned into CSE's:

Examples are `CORINFO_HELP_GETSHARED_GCSTATIC_BASE` and  `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE`

Comment (mikedn) [reply src/jit/optimizer.cpp]: Those are already handled in `HoistVisitor::PostOrderVisit` - it already checks for pure helpers with invariant args.

As far as I can tell the only good reason to check VN invariance is to detect memory loads that are invariant. Apart from that the VN invariance check is effectively redundant and confusing:
* `PostOrderVisit` determines that `a` and `b` are invariant and then trivially concludes that `a + b` is also invariant.
* At that point it checks VN invariance - it goes "down" the VN "tree" to check if `a + b` is invariant

Basically as `PostOrderVisit` goes up the expression tree it keeps going down the VN tree again and again attempting to check what has already been checked. If it weren't for `loopVnInvariantCache` this would probably rather slow.

To make things worse, VN invariance could handle the case of a variable defined inside the loop by an invariant expression. Sadly this doesn't happen because `PostOrderVisit` specifically requires that the definition be outside the loop and doesn't check VN invariance if the def is inside the loop.

**Comment by briansull** [comment]
No diffs in --frameworks

File: src/jit/optimizer.cpp [hunk 4]
```diff
@@ -6795,8 +6795,22 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
...
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // We only allow true constant nodes  (i.e. GT_CNS_INT, etc...) or a GT_ARGPLACE nodes.
+                // It is unsafe to allow a GT_CLS_VAR that has been assigned a constant.
+                // The logic in optVNIsLoopInvariant woudl consider it to be loop-invariant, even
```
Comment (erozenfeld) [inline src/jit/optimizer.cpp]: Typo: woudl --> would

File: src/jit/optimizer.cpp [hunk 5]
```diff
@@ -6795,8 +6795,22 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
 
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // We only allow true constant nodes  (i.e. GT_CNS_INT, etc...) or a GT_ARGPLACE nodes.
```
Comment (CarolEidt) [inline src/jit/optimizer.cpp]: This should be "... or GT_ARGPLACE nodes" (omit the "a")

File: src/jit/optimizer.cpp [hunk 6]
```diff
@@ -6795,8 +6795,22 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
...
+                // It is unsafe to allow a GT_CLS_VAR that has been assigned a constant.
+                // The logic in optVNIsLoopInvariant woudl consider it to be loop-invariant, even
+                // if the assignment of the constant to the GT_CLS_VAR was inside the loop.
+                //
+                if (!tree->OperIsConst() && tree->OperGet() != GT_ARGPLACE)
```
Comment (CarolEidt) [inline src/jit/optimizer.cpp]: This would be cleaner as 
```
if (!tree->OperIsConst() && !tree->OperIs(GT_ARGPLACE))
```
But if using the inequality, it should be parenthesized.

**Comment by briansull** [comment]
With the current fix we will see a regression in a couple of the benchmarks:
         -39 : JIT\Performance\CodeQuality\Benchstones\BenchF\Simpsn\Simpsn\Simpsn.dasm (-4.74% of base)
         -29 : JIT\Performance\CodeQuality\Benchstones\BenchF\Trap\Trap\Trap.dasm (-3.66% of base)


The old implementation will hoist expressions such as this one found in 
`Performance\CodeQuality\Benchstones\BenchF\Simpsn`

```
Hoisting a copy of [000223] into PreHeader for loop L01 <BB03..BB05>:
N005 ( 46, 15) [000223] --C---------              *  INTRINSIC double exp $285
N004 ( 10, 11) [000222] ------------              \--*  MUL       double $49
N002 (  2,  3) [000220] ------------                 +--*  NEG       double $49
N001 (  1,  2) [000098] ------------                 |  \--*  LCL_VAR   double V00 loc0         u:2 $40
N003 (  3,  4) [000221] ------------                 \--*  CNS_DBL   double 2.0000000000000000 $45

```

**Comment by briansull** [comment]
I will implement a new fix that replaces the constant value numbers that are not represent by constants with true constants such as GT_CNS_INT or GT_CNS_DBL

**Comment by briansull** [comment]
I am going to limit this to just GT_CLS_VARS. so that we don't regress the cases that improve the CodeQuality tests.

**Comment by erozenfeld** [comment]
With this fix we still have a bug here since we'll hoist any non-GT_CLS_VAR node with a constant value number, which is unsafe. I like your previous suggestion 
>>> I will implement a new fix that replaces nodes in the hoisted expressions that have constant value number, but are not true constants such as GT_CNS_INT or GT_CNS_DBL

better. Was there a problem with that approach?

**Comment by briansull** [comment]
It is a much bigger change and I don't  think it is required to fix this issue.
Only assignments can create constant value-numbers.
I will continue to investigate further, and likely implement the other fix.
But I would prefer to close this issue with a  small fix that has zero Asm Diffs.

**Comment by erozenfeld** [comment]
The issue can potentially happen with assignments of constants to non-faulting indirections (e.g., fields or array elements). I couldn't get a repro with an indirection but only because we recognize the indirections as non-faulting too late.

**Comment by jkotas** [comment]
@briansull This PR was merged with failing test in `coreclr-ci (Run Test Pri0 Windows_NT x86 checked).`. This failure was most likely introduced by this change and all PR jobs are failing because of it. I submitting a PR to revert this change.

**Comment by BruceForstall** [comment]
New proposed fix: https://github.com/dotnet/coreclr/pull/26952
